### PR TITLE
Implementation for a Network-Wide Stop Command

### PR DESCRIPTION
**Pull request by Zorro909**
This pull Request implements the StopCommand with a Protocol Addition of the Message Type "SHUTDOWN" which tells the Server to shutdown immediately.
In addition the Servers now won't shutdown due to a loss of connection to the core and instead will try to reconnect afterwards. Since a hard exit is now not needed to stop the system, the servers won't die just because the Core died.

### CODE REVIEW

**Comment by TimoZikeli** [comment]
What about the Core‘s current state? The Core requires the list of running servers/proxies in order to work properly

File: TimoCloud-Universal/src/main/java/cloud/timo/TimoCloud/bukkit/sockets/BukkitStringHandler.java [hunk 1]
```diff
@@ -52,6 +52,12 @@ public void handleMessage(Message message, String originalMessage, Channel chann
                 ((TimoCloudMessageAPIBasicImplementation) TimoCloudAPI.getMessageAPI()).onMessage(addressedPluginMessage);
                 break;
             }
+            case "SHUTDOWN":
+            	if(TimoCloudBukkit.getInstance().getServerName().equalsIgnoreCase((String) message.get("name"))) {
```
Comment (TimoZikeli) [inline TimoCloud-Universal/src/main/java/cloud/timo/TimoCloud/bukkit/sockets/BukkitStringHandler.java:56]: Shouldn‘t the server just get shutodwn messages which are addressed to it?

File: TimoCloud-Universal/src/main/java/cloud/timo/TimoCloud/bungeecord/TimoCloudBungee.java [hunk 2]
```diff
@@ -27,227 +35,231 @@
...
-
-    public boolean isShuttingDown() {
-        return shuttingDown;
-    }
-
-    public void setShuttingDown(boolean shuttingDown) {
-        this.shuttingDown = shuttingDown;
-    }
-
+	private static TimoCloudBungee instance;
```
Comment (TimoZikeli) [inline TimoCloud-Universal/src/main/java/cloud/timo/TimoCloud/bungeecord/TimoCloudBungee.java:40]: Please use the TimoCloud default indentation.

File: TimoCloud-Universal/src/main/java/cloud/timo/TimoCloud/bungeecord/sockets/BungeeStringHandler.java [hunk 3]
```diff
@@ -73,6 +73,13 @@ public void handleMessage(Message message, String originalMessage, Channel chann
                 ((TimoCloudMessageAPIBasicImplementation) TimoCloudAPI.getMessageAPI()).onMessage(addressedPluginMessage);
                 break;
             }
+            case "SHUTDOWN":
+            	if(TimoCloudBungee.getInstance().getProxyName().equalsIgnoreCase((String) message.get("name"))) {
```
Comment (TimoZikeli) [inline TimoCloud-Universal/src/main/java/cloud/timo/TimoCloud/bungeecord/sockets/BungeeStringHandler.java:77]: Shouldn‘t the proxy just get shutdown messages which are addressed to it?

File: TimoCloud-Universal/src/main/java/cloud/timo/TimoCloud/core/objects/Proxy.java [hunk 4]
```diff
@@ -112,7 +112,8 @@ public void start() {
 
     @Override
     public void stop() {
-        if (getChannel() != null) getChannel().close();
+    	getBase().sendMessage(Message.create().setType("SHUTDOWN").set("name", getName()));
```
Comment (TimoZikeli) [inline TimoCloud-Universal/src/main/java/cloud/timo/TimoCloud/core/objects/Proxy.java:115]: Why are you sending the message to the base?

**Comment by TimoZikeli** [comment]
What is this PR‘s goal? Should it avoid accidental stopping of the core or make updating easier?

**Comment by Zorro909** [comment]
This PR should make shutting down the whole Network easier, since shutting down the Core leaves the other parts running. On the other hand if a error occurs in the Core and it crashes or sth. similar happens, all MC Servers and such would shut down immediately, which disrupts the Player Experience. And about the currently running servers and proxies, they reconnect immediately after the Core started again, so it should be able to recognize that they are online or did i miss something on that part?

PS: I'm working together with a Server Team that requested this feature to be in a CloudSystem that they would want to use. And i thought it could also help others, so if you have any suggestions as to how it could be handled better, i'm completely open for that.

**Comment by TimoZikeli** [comment]
Have you tested it? Servers/proxies won‘t be able to reconnect properly, since they aren‘t registered in the CoreInstanceManager and their token will be rejected when they want to do the handshake.

I think it‘s impossible for the Core to crash. Even if it did, the whole network wouldn‘t work anymore (synchronisation). Stopping the whole network can also be done simpler: Just leave the system as it currently is, and add a shutdown command to stop bases and cords.

**Comment by Zorro909** [comment]
What exactly was the point behind the token-based session system?

**Comment by Zorro909** [comment]
Since i don't think this leads to anywhere, i'm gonna use this project as a base and develop it further as a fork. 
Thanks

**Comment by TimoZikeli** [comment]
Sorry, since I was on vacation, I didn‘t see your question about the token authentication system. It‘s all about security - in order to prevent foreign servers to connect and control your network via the API.

Please don‘t give up - I appreciate your pull request! Though, I think a lot about features before implementing them. Hence, I would enjoy discussing your PR with you and implement the resulting solution instead of leading you to go your own way.

**Comment by Zorro909** [comment]
Oh, okay, then how about we meet on the discord server and try to figure everything out? The Owner of the Server i am working with and I have a lot of features/things in mind that we would like to optimize, so i would really appreciate if we could work together on that.

**Review by base2code — APPROVED** [review]
Everything is okay

**Review by TimoZikeli — CHANGES_REQUESTED** [review]

**Comment by TimoZikeli** [comment]
Apart from the requested changes, some of the used methods are already outdated by now :)
In addition, I strongly believe that this PR has not been tested because it is impossible that the servers reconnect after the Core has been restarted because the Core's state is not being saved.

Hence, I am closing it for now.

**Pull request closed without merging**